codalab/apps/api/views/competition_views.py

```diff
@@ -646,11 +645,92 @@ def addToLeaderboard(self, request, pk=None, competition_id=None):
 competition_submission_leaderboard = CompetitionSubmissionViewSet.as_view(
                                         {'post':'addToLeaderboard', 'delete':'removeFromLeaderboard'})
 
+
+class CompetitionSubmissionListViewSet(mixins.ListModelMixin, viewsets.GenericViewSet):
+    queryset = CompetitionSubmission.objects.all()
+    serializer_class = serializers.CompetitionSubmissionListSerializer
+    # paginate_by = 10
+    # paginate_by_param = 'page_size'
+
+    def get_queryset(self, *args, **kwargs):
+        qs = super(CompetitionSubmissionListViewSet, self).get_queryset(*args, **kwargs)
+
+        # Only get submissions for this competition, and only if you're an admin
+        competition_id = self.kwargs['competition_id']
+        qs = qs.filter(phase__competition_id=competition_id).order_by('-pk')
+        qs = qs.filter(Q(phase__competition__creator=self.request.user) | Q(phase__competition__admins__in=[self.request.user]))
+
+        qs = qs.extra(
+            select={
+                'participant_submission_number':
+                    # 'DISTINCT "web_competitionsubmission"."participant_id" '
+                    # 'DENSE_RANK() OVER(ORDER BY "web_competitionsubmission"."id" ASC) '
+                    # 'DENSE_RANK() OVER(ORDER BY "web_competitionsubmission"."id" ASC) '
+                    'DENSE_RANK() OVER(PARTITION BY "web_competitionsubmission"."participant_id" ORDER BY "web_competitionsubmission"."id" ASC) '
+                    # 'GROUP BY "web_competitionsubmission"."participant_id"'
+            }
+        )
+
+        qs = qs.select_related(
+            'status',
+            'participant',
+            'participant__user',
+            'phase',
+            'phase__competition',
+        )
+        return qs
+
+    def get_serializer_context(self, *args, **kwargs):
+        context = super(CompetitionSubmissionListViewSet, self).get_serializer_context(*args, **kwargs)
+
+        # To reduce queries, let's collect a bit of data to make processing submissions easier
+        competition_id = self.kwargs['competition_id']
+        context['leaderboard_submissions'] = PhaseLeaderBoardEntry.objects.filter(
+            board__phase__competition_id=competition_id
+        ).values_list('result__id', flat=True)
+
+        first_phase, previous_phase, active_phase, next_phase = get_first_previous_active_and_next_phases(
+            Competition.objects.get(pk=competition_id)
+        )
+
+        # Get all submissions that can be migrated into another phase (must be on leaderboard already and
+        # the next phase must have auto_migration
+        if next_phase and next_phase.auto_migration:
+            active_phase_submissions = CompetitionSubmission.objects.filter(
+                phase=active_phase,
+            ).values_list('id', flat=True)
+            context['migratable_submissions'] = filter(lambda x: x in context["leaderboard_submissions"], active_phase_submissions)
```

Comment:
Would a list comprehension work here?

Py3 seems to prefer list comprehension over `filter(lambda)`

IE:
```
context['migratable_submissions'] = [submission for submission in context["leaderboard_submissions"] if submission in active_phase_submissions]
```

It looks like list comprehensions might be faster than filter(lambda) as well.